sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/CSVExprUtils.scala

```diff
@@ -21,38 +21,34 @@ import org.apache.commons.lang3.StringUtils
 
 object CSVExprUtils {
   /**
-   * Filter ignorable rows for CSV iterator (lines empty and starting with `comment`).
+   * Filter ignorable rows starting with `comment`.
    * This is currently being used in CSV reading path and CSV schema inference.
    */
-  def filterCommentAndEmpty(iter: Iterator[String], options: CSVOptions): Iterator[String] = {
+  def filterComment(iter: Iterator[String], options: CSVOptions): Iterator[String] = {
     if (options.isCommentSet) {
       val commentPrefix = options.comment.toString
-      iter.filter { line =>
-        line.trim.nonEmpty && !line.startsWith(commentPrefix)
-      }
+      iter.dropWhile(_.startsWith(commentPrefix)
     } else {
-      iter.filter(_.trim.nonEmpty)
```

Comment:
IIRC this matters when you call `DataFrameReader.csv(dataset: Dataset[String])`

Reply:
You may be correct. However, it seems the actual read path of `DataFrameReader.csv(dataset: Dataset[String])` goes through 'CSVUtils.filterCommentAndEmpty(lines: Dataset[String], options: CSVOptions)' whereas `TextInputCSVDataSource.inferFromDataset(...)` goes through `CSVUtils.filterCommentAndEmpty(iter: Iterator[String], options: CSVOptions)` (which calls `CSVExprUtils.filterCommentAndEmpty(iter: Iterator[String], options: CSVOptions)`. In any case, I think there is definitely an oportunity to make things a little more clear here and consistent between multiline and non-multiline (though different for `Dataset[String]`) If i figure something out, I'll open a ticket. Will close this PR if not.

Reply:
I've addressed this now, things a bit more consistent. Only `CSVUtils` is concerned with the `Dataset[String]` case and the CSV file reads are concerned with `CSVExprUtils`. Would be interested to see what you think of this as I was very confused when trying to do some CSV work to find blank lines were not in fact removed here.